src/runtime/typemanager.cs

```diff
@@ -817,4 +963,183 @@ internal static void CopySlot(IntPtr from, IntPtr to, int offset)
             Marshal.WriteIntPtr(to, offset, fp);
         }
     }
+
+    class SlotsHolder
+    {
+        public const string HolderKeyName = "_slots_holder";
+        public delegate void Resetor(IntPtr type, int offset);
+
+        private GCHandle _handle;
+        private Interop.DestructorFunc _destructor;
+        private IntPtr _capsule;
+        private IntPtr _type;
+        private Dictionary<int, ThunkInfo> _slots = new Dictionary<int, ThunkInfo>();
+        private List<Delegate> _keepalive = new List<Delegate>();
+        private Dictionary<int, Resetor> _customRestors = new Dictionary<int, Resetor>();
+        private List<Action> _deallocators = new List<Action>();
+        private bool _alredyReset = false;
+
+        /// <summary>
+        /// Create slots holder for holding the delegate of slots and be able  to reset them.
+        /// </summary>
+        /// <param name="type">Steals a reference to target type</param>
+        public SlotsHolder(IntPtr type)
+        {
+            _type = type;
+        }
+
+        public void Add(int offset, ThunkInfo thunk)
+        {
+            _slots.Add(offset, thunk);
+        }
+
+        public void Add(int offset, Resetor resetor)
+        {
+            _customRestors[offset] = resetor;
+        }
+
+        public void AddDealloctor(Action deallocate)
+        {
+            _deallocators.Add(deallocate);
+        }
+
+        public void KeeapAlive(Delegate d)
+        {
+            _keepalive.Add(d);
+        }
+
+        public IntPtr ToCapsule()
+        {
+            if (_capsule != IntPtr.Zero)
+            {
+                Runtime.XIncref(_capsule);
+                return _capsule;
+            }
+            _handle = GCHandle.Alloc(this);
+            _destructor = OnDestruct;
+            var fp = Marshal.GetFunctionPointerForDelegate(_destructor);
+            _capsule = Runtime.PyCapsule_New((IntPtr)_handle, null, fp);
+            return _capsule;
+        }
```

Comment:
From a name `ToCapsule`, I would expect this method to behave similarly as `ToString`. E.g. first, I'd expect it to not modify the object state at all, second, I'd expect it to construct a new object every time it is called, and for it to be callable multiple times.

Overall, it would be good to either have a more detailed description of what this class does, and how capsules fit into that, or refactor code to it would be clear what it does.

Reply:
It is also unclear why you can't just use `slotsHolder.ToPython()`.

Reply:
>[...] I would expect this method to behave similarly as ToString
I agree, I'll change it.

From what I see the usage of the capsule fits the semantics they were created for. Maybe we  could also use a regular. @amos402 Any other reason to use a capsule that I'm not seeing?

Reply:
> [...] I would expect this method to behave similarly as ToString

I agree, I'll change it.

From what I see the usage of the capsule fits the semantics they were created for. Maybe we could also use a regular. @amos402 Any other reason to use a capsule that I'm not seeing?

Reply:
`ToPython` cannot be use here, some functions of pythonnet had been shutdown.
I used capsule as a simple mechanism to track the life cycle of type, so it can call a callback when then type deallocated, we can handle this manually as well if my conception correct. So my next step would be try to remove this mechanism.